apps/aetx/src/aetx_env.erl

```diff
@@ -103,13 +88,39 @@ tx_env_from_key_header(KeyHeader, KeyHash, Time,_PrevHash) ->
         , time = Time
         }.
 
+tx_env_and_trees_from_top(Type) when Type == aetx_contract;
+                                     Type == aetx_transaction ->
+    {TopHeader, TopHash, Trees} = aec_chain:top_header_hash_and_state(),
+    {KeyHeader, KeyHash, Time} =
+        case aec_headers:type(TopHeader) of
+            micro ->
+                KHash = aec_headers:prev_key_hash(TopHeader),
+                {ok, KH} = aec_chain:get_header(KHash),
+                {KH, KHash, aec_headers:time_in_msecs(TopHeader)};
+            key ->
+                {TopHeader, TopHash, aec_headers:time_in_msecs(TopHeader) + 1}
```

Comment:
I see the intention on time here is to simulate the first legal timestamp for the first microblock. I see same snippet `aec_blocks:time_in_msecs(LastKeyBlock) + 1` for [micro block candidate](https://github.com/aeternity/epoch/blob/fce48d3542f7cace2d0ccba8c9f84f2f303dbe97/apps/aecore/src/aec_block_micro_candidate.erl#L144). Factoring this in common would help readability. Though totally optional of course.